### PR TITLE
feat: implement k anonymity (#45)

### PR DESCRIPTION
**Pull request by stellaljung**

### CODE REVIEW

File: backend/src/server.cpp [hunk 1]
```diff
@@ -31,7 +31,14 @@ namespace server
         std::function <std::string(sqlite3_stmt *)> callback = [](sqlite3_stmt *stmt) {
             return std::string(reinterpret_cast<const char *>(sqlite3_column_text(stmt, 0)));
         };
-        std::vector<std::string> breached_passwords = db.execute("SELECT * FROM passwords;", callback);
+
+        // get the table num corresponding to the user password leaked byte
+        std::string decoded_password = crow::utility::base64decode(user_password, user_password.size());
```
Comment (ni-jessica) [inline backend/src/server.cpp:36]: **nitpick**: could we use `decoded_password` in the `encryptPassword` function below?

File: backend/src/main.cpp [hunk 2]
```diff
@@ -67,31 +79,39 @@ int main(int argc, char *argv[])
...
         {
             int count = atoi(reinterpret_cast<const char *>(sqlite3_column_text(stmt, 0)));
             return count;
         };
 
-        // check if passwords and key table exists
-        std::vector<bool> password_result = db.execute("SELECT COUNT(*) FROM sqlite_schema WHERE name = 'passwords';", callback);
-        std::vector<bool> secret_result = db.execute("SELECT COUNT(*) FROM sqlite_schema WHERE name = 'secret';", callback);
-        if (password_result.front() == 0 || secret_result.front() == 0) // no passwords or no key table exists
+        // check if key ad offset table exists
```
Comment (ni-jessica) [inline backend/src/main.cpp:109]: **nitpick:** and*

**Review by ni-jessica — APPROVED** [review]

File: backend/src/main.cpp [hunk 3]
```diff
@@ -47,10 +47,22 @@ int main(int argc, char *argv[])
...
     if (build)
     {
-        // create password table
-        db.execute("CREATE TABLE passwords (password TEXT);");
+        // create all tables from 0 to (2^8)^offset-1
+        for (int i = 0; i < std::pow(std::pow(2, 8), offset); i++)
+        {
+            std::string query_str = "table" + std::to_string(i);
+            db.execute("CREATE TABLE " + query_str + " (password TEXT);");
+            spdlog::info("Created table: {}", query_str);
```
Comment (csirianni) [inline backend/src/main.cpp:64]: **suggestion**: This is too much logging for `info`. Could be `debug` or just removed.

File: backend/src/main.cpp [hunk 4]
```diff
@@ -67,31 +79,39 @@ int main(int argc, char *argv[])
         // 3. insert into database
         for (const auto &password : encrypted_passwords)
         {
+            // determine which table to insert into based on leaked byte
+            std::string encoded_byte = crow::utility::base64encode(password.substr(0, offset), offset);
+            std::string table_num = "table" + std::to_string(static_cast<unsigned int>(encoded_byte[0]));
```
[the remark this answers is not in the document]
Comment (ni-jessica) [reply backend/src/main.cpp:84]: We tried to use ints and it was erroring

File: backend/tests/server.cpp [hunk 5]
```diff
@@ -73,38 +93,27 @@ TEST_CASE("Test endpoints using handler")
...
-                CHECK(breached_password[breached_password.size() - 2] == '=');
-                break;
-            }
-        }
+        std::vector<std::string> breached_bucket = body["breachedPasswords"].get<std::vector<std::string>>();
+        CHECK(breached_bucket.size() == 1);
+
+        // check breached password encoding
+        std::string breached_password = breached_bucket[0];
+        CHECK(breached_password[breached_password.size() - 1] == '=');
```
[the remark this answers is not in the document]
Comment (ni-jessica) [reply backend/tests/server.cpp:114]: We store the raw password value so now they should always be 32 bytes

File: backend/src/main.cpp [hunk 6]
```diff
@@ -67,31 +79,39 @@ int main(int argc, char *argv[])
...
+
+            std::string raw_password = password.substr(offset, password.size() - offset);
```
[the remark this answers is not in the document]
Comment (ni-jessica) [reply backend/src/main.cpp:86]: this returns password[offset:]. we could maybe use a sodium constant like crypto_core_ristretto255_BYTES for readability

File: backend/src/main.cpp [hunk 7]
```diff
@@ -67,31 +79,39 @@ int main(int argc, char *argv[])
         // 3. insert into database
         for (const auto &password : encrypted_passwords)
         {
+            // determine which table to insert into based on leaked byte
+            std::string encoded_byte = crow::utility::base64encode(password.substr(0, offset), offset);
```
[the remark this answers is not in the document]
Comment (ni-jessica) [reply backend/src/main.cpp:83]: When we don't encode the byte, the int conversion sometimes causes the leaked byte to be casted to a negative number. We tried to use unsigned int to solve the issue but this caused some UB, so we chose to encode and then cast to int

**Review by ni-jessica — DISMISSED** [review]

**Review by csirianni — APPROVED** [review]

**Review by ni-jessica — APPROVED** [review]

**Pull request merged**